### PR TITLE
Add support for named arguments to sprintf.

### PR DESCRIPTION
**Pull request by sgomes**
This allows for `sprintf` usage of the type:
```js
sprintf( '%(a)s %(b)s', { a: 'foo', b: 'bar' } );
```

While not part of the standard GNU `printf` feature set, named arguments are supported by several `printf` implementations, notably `Jed`'s.

### CODE REVIEW

File: packages/sprintf/index.js [hunk 1]
```diff
@@ -67,14 +68,20 @@ export default function sprintf( string, args ) {
 			i++;
 		}
 
-		// If not a positional argument, use counter value.
-		if ( index === undefined ) {
-			index = i;
-		}
+		if ( name !== undefined ) {
+			// If it's a named argument, use name.
+			value = args[ 0 ] && args[ 0 ][ name ];
```
Comment (aduth) [inline packages/sprintf/index.js]: I haven't verified, but I think the use of `&&` to assume short-circuit evaluation on `args[ 0 ]` could result in some odd behavior here, since it will defer to the value of the left-hand of the condition when falsey.

Specifically, I might wonder if this could result:

```js
sprintf( 'My name is %(name)s', 0 )
// "My name is 0"
```

Or another odd one, caused by object prototype inheritance:

```js
sprintf( 'My name is %(valueOf)s', {} );
// "My name is function valueOf() { [native code] }"
```

I might consider instead:

```suggestion
			value = args[ 0 ] && args[ 0 ].hasOwnProperty( name ) ? args[ 0 ][ name ] : '';
```

Or the more verbose `if` condition, to avoid long lines:


```suggestion
			if ( args[ 0 ] && args[ 0 ].hasOwnProperty( name ) ) {
				value = args[ 0 ];
			} else {
				value = '';
			}
```

Comment (aduth) [reply packages/sprintf/index.js]: With #1, we could consider the `if` condition and omit the `else` case, since it would leave `value` as `undefined` and implicitly assign an empty string result.

File: packages/sprintf/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,9 @@
+## 1.1.0 (2019-07-25)
```
Comment (aduth) [inline packages/sprintf/CHANGELOG.md]: Both here and in #2, can you leave the date as unset? I'm not sure I can commit to having this published today.

```suggestion
## 1.1.0 (TBD)
```

File: packages/sprintf/index.js (hunk 1 above)
Comment (sgomes) [reply packages/sprintf/index.js]: The first one is not an issue, but the second one is. Went with the second approach.

File: packages/sprintf/CHANGELOG.md (hunk 2 above)
Comment (sgomes) [reply packages/sprintf/CHANGELOG.md]: Can do!

File: packages/sprintf/index.js (hunk 1 above)
Comment (aduth) [reply packages/sprintf/index.js]: Can we add tests for these edge cases?

I thought there might be another one:

```js
sprintf( 'My name is %(name)s', 1 )
```

...where I expected `hasOwnProperty` to not be available for `1`, despite truthy. But it works.

I guess because every of these inherits from `Object` 🤔 

```js
( 1 ).constructor instanceof Object
// true
```

Comment (sgomes) [reply packages/sprintf/index.js]: Good idea! I added these, plus a few more I could think of. One of them was failing, so I fixed it too 🙂
